server/src/main/java/org/candlepin/model/CandlepinQuery.java

```diff
@@ -0,0 +1,254 @@
+ * executing criteria, and methods for processing the results.
+ *
+ * @param <T>
+ *  The entity type to be returned by this criteria's result output methods
+ */
+public interface CandlepinQuery<T> extends Iterable<T> {
+
+    // TODO:
+    // Add support for stateless sessions (which requires some workarounds because stateless sessions
+    // and sessions don't have a common parent class)
+
+    /**
+     * Sets the session to be used for executing this query
+     *
+     * @param session
+     *  The session to use for executing this query
+     *
+     * @throws IllegalArgumentException
+     *  if session is null
+     *
+     * @return
+     *  this query instance
+     */
+    CandlepinQuery<T> useSession(Session session);
+
+    /**
+     * Sets the offset (first result) into a result set at which to begin fetching results.
+     *
+     * @param offset
+     *  The offset at which to begin fetching results when executing this query. Negative values
+     *  will clear any previously set offset.
+     *
+     * @return
+     *  this query instance
+     */
+    CandlepinQuery<T> setFirstResult(int offset);
+
+    /**
+     * Sets the maximum results to be returned when executing this query.
+     *
+     * @param limit
+     *  The maximum number of results to be returned when executing this query. Negative values
+     *  will disable any previously set limits.
+     *
+     * @return
+     *  this query instance
+     */
+    CandlepinQuery<T> setMaxResults(int limit);
+
+    /**
+     * Adds the specified ordering when executing this query.
+     *
+     * @param order
+     *  The ordering to apply when executing this query
+     *
+     * @return
+     *  this query instance
+     */
+    CandlepinQuery<T> addOrder(Order order);
+
+    // TODO:
+    // Add some other utility/passthrough methods as a need arises:
+    //  - setReadOnly
+    //  - setFetchMode
+    //  - setCacheMode/setCacheable
+
+    /**
+     * Executes this query and returns the entities as a list. If no entities could be found,
+     * this method returns an empty list.
+     * <p></p>
+     * <strong>Warning</strong>:
+     * This method loads the entire result set into memory. As such, this method should not be used
+     * with queries that can return extremely large data sets.
+     *
+     * @return
+     *  a list containing the results of executing this criteria
+     */
+    List<T> list();
+
+    /**
+     * Steps through the results of a column of the given query row-by-row, rather than dumping the
+     * entire query result into memory before processing it. This method will always pass the first
+     * column of each row to the processor.
+     *
+     * @param processor
+     *  A ResultProcessor instance to use for processing each result
+     *
+     * @return
+     *  the number of rows processed and sent to the result processor
+     */
+    int scroll(ResultProcessor<T> processor);
```

Comment:
Since Java 8 uses `forEach` as the method name with the same semantics, I think we should rename `scroll` to `forEach` for consistency.